mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,724 @@
+
+        Returns:
+            file_entities_map (dict): A dictionary that maps file paths in an
+                App to a list of entities.
+        """
+        all_file_paths = self._resource_loader.get_all_file_paths()
+        file_entities_map = {path: [] for path in all_file_paths}
+
+        if action == AnnotatorAction.ANNOTATE:
+            rules = self.config[AnnotatorAction.ANNOTATE.value]
+        elif action == AnnotatorAction.UNANNOTATE:
+            rules = self.config[AnnotatorAction.UNANNOTATE.value]
+
+        for rule in rules:
+            pattern = Annotator._get_pattern(rule)
+            compiled_pattern = re.compile(pattern)
+            filtered_paths = self._resource_loader.filter_file_paths(
+                compiled_pattern=compiled_pattern, file_paths=all_file_paths
+            )
+            for path in filtered_paths:
+                entities = self._get_entities(rule)
+                file_entities_map[path] = entities
+        return file_entities_map
+
+    @staticmethod
+    def _get_pattern(rule):
+        """ Convert a rule represented as a dictionary with the keys "domains", "intents",
+        "entities" into a regex pattern.
+
+        Args:
+            rule (dict): Annotation/Unannotation rule.
+
+        Returns:
+            pattern (str): Regex pattern specifying allowed file paths.
+        """
+        pattern = [rule[x] for x in ["domains", "intents", "files"]]
+        return ".*/" + "/".join(pattern)
+
+    def _get_entities(self, rule):
+        """ Process the entities specified in a rule dictionary. Check if they are valid
+        for the given annotator.
+
+        Args:
+            rule (dict): Annotation/Unannotation rule with an "entities" key.
+
+        Returns:
+            valid_entities (list): List of valid entities specified in the rule.
+        """
+        if rule["entities"].strip() in ["*", ".*", ".+"]:
+            return ["*"]
+        entities = re.sub("[()]", "", rule["entities"]).split("|")
+        valid_entities = []
+        for entity in entities:
+            entity = entity.strip()
+            if self.valid_entity_check(entity):
+                valid_entities.append(entity)
+            else:
+                logger.warning("%s is not a valid entity. Skipping entity.", entity)
+        return valid_entities
+
+    @abstractmethod
+    def valid_entity_check(self, entity):
+        """ Determine if an entity type is valid.
+
+        Args:
+            entity (str): Name of entity to annotate.
+
+        Returns:
+            bool: Whether entity is valid.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+    def annotate(self, **kwargs):
+        """ Annotate data based on configurations in the config.py file.
+
+        Args:
+            kwargs (dict, optional): Configuration overrides can be passed in as arguments.
+        """
+        for key, value in kwargs.items():
```

Comment:
we probably do not want to permanently change the config, maybe just overriding the config for this operation

Reply:
Without modifying the base config, I would need to pass in the current config downstream to each downstream function as these functions get configuration details from the self attribute and not from a function parameter.

Should I add a "config" param for the downstream functions in order to use an alternate config?

Reply:
Yup let's do that instead